### PR TITLE
Use DistinctSorted only when applicable

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use `DistinctSortedTransform` only when sort description is applicable to DISTINCT columns, otherwise fall back to ordinary DISTINCT implementation. It allows making less checks during `DistinctSortedTransform` execution

### CODE REVIEW

File: src/Processors/QueryPlan/DistinctStep.cpp [hunk 1]
```diff
@@ -140,15 +140,19 @@ void DistinctStep::transformPipeline(QueryPipelineBuilder & pipeline, const Buil
...
-                    pipeline.addSimpleTransform(
-                        [&](const Block & header, QueryPipelineBuilder::StreamType stream_type) -> ProcessorPtr
-                        {
-                            if (stream_type != QueryPipelineBuilder::StreamType::Main)
-                                return nullptr;
-
-                            return std::make_shared<DistinctSortedTransform>(
-                                header, distinct_sort_desc, set_size_limits, limit_hint, columns);
-                        });
+                    if (DistinctSortedTransform::isApplicable(pipeline.getHeader(), distinct_sort_desc, columns))
```
Comment (devcrafter) [inline src/Processors/QueryPlan/DistinctStep.cpp:143]: Add tests with EXPLAIN to check that `DistinctSorted` is not used in non-applicable cases

**Comment by devcrafter** [comment]
@nickitat @KochetovNicolai would you mind to review?

**Comment by devcrafter** [comment]
@CurtizJ please could you have a look?

**Review by CurtizJ — APPROVED** [review]
LGTM

**Comment by CurtizJ** [comment]
@devcrafter 
Changelog entry should be updated, because current description is not user-readable.

**Pull request merged**